### PR TITLE
allow stands to auto restore

### LINKED ISSUE #16: Track the number of times a mocked method is called
Just thought this might be useful. Motivation comes from https://github.com/hapijs/code/issues/66.

### LINKED ISSUE #17: Allow stand to auto restore itself
This is related to #16 a bit, but it would be nice if a stand could auto-restore itself after being called `n` times.

### LINKED ISSUE #18: Add option to specify when original or monkey patched version is run
It would be nice if you could say you want the stand in version to run on invocations 1, 3, 4, etc. And in other cases, the original function would execute.

### PR DESCRIPTION
**Pull request by cjihrig**
Closes #16
Closes #17
Closes #18

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -25,31 +26,56 @@ internals.store.lookUp = function (obj, path) {
...
+
+    var args = new Array(arguments.length + 1);
+
+    args[0] = result;
+
+    for (var i = 0; i < arguments.length; ++i) {
+      args[i + 1] = arguments[i];
+    }
+
+    return fn.apply(obj, args);
```
Comment (mattlgy) [inline lib/index.js]: This line would blow away `this` for things with `new`, I think?

**Comment by arb** [comment]
Needs doc updates, looks good otherwise though.

File: lib/index.js [hunk 2]
```diff
@@ -2,6 +2,7 @@
 // Load modules
 var Assert = require('assert');
 var Get = require('lodash.get');
+var Merge = require('lodash.merge');
```
Comment (arb) [inline lib/index.js:5]: Since options is only one level deep, did we really need to add this dep?

File: lib/index.js [hunk 3]
```diff
@@ -25,31 +26,56 @@ internals.store.lookUp = function (obj, path) {
...
+
   var result = {
     original: func,
-    restore: function () {
-      var find = internals.store.lookUp(obj, path);
-      internals.store.splice(find.index, 1);
-      Set(obj, path, this.original);
-    },
+    restore: restore,
+    get invocations () { return invocations; },
```
Comment (arb) [inline lib/index.js:58]: Why a `get` here we never really use that pattern do we?

File: lib/index.js [hunk 4]
```diff
@@ -25,31 +26,56 @@ internals.store.lookUp = function (obj, path) {
...
     _stand: stand
   };
 
-  Set(obj, path, fn.bind(obj, result));
+  function wrapper () {
+    invocations++;
+
+    if (invocations === options.restoreAfter) {
```
Comment (arb) [inline lib/index.js]: This is a little wordy of an option. `ttl` or `tte` maybe? "time to live" or "times to execute"? Maybe just `count`? I'd suggest `invocations`, but I'd never be able to spell it right.

File: lib/index.js (hunk 2 above)
Comment (cjihrig) [inline lib/index.js:5]: I originally used `Object.assign()` but it made the Node < 4 CI blow up :-)

File: lib/index.js (hunk 3 above)
Comment (cjihrig) [inline lib/index.js:58]: To expose `invocations` but keep the outside world from changing its value.

File: lib/index.js (hunk 4 above)
Comment (cjihrig) [inline lib/index.js]: I don't like `ttl`. It implies a passage of an amount of time, which is completely unrelated to this. I'd like to use a name that has a well defined meaning or something that actually describes what it does. I couldn't come up with a well defined term.

Comment (arb) [inline lib/index.js]: I don't really like the long name... `cycles` maybe? `maxCycles`?

**Pull request merged**